### PR TITLE
Add RFC7233 support

### PR DESCRIPTION
**Pull request by denysvitali**
This PR adds support for RFC7233 (multipart data), aka file streaming.
~~I don't like my "hack" to extend the Response class, so that may need some cleanup, if someone comes with a better / cleaner idea I'll be happy to fix my PR.~~
Nevermind, fixed in commit 83475c5

Anyways, I've tested it on my machine and it works: this should finally bring file streaming to Kemal.

### CODE REVIEW

**Comment by denysvitali** [comment]
This btw fixes #276

**Comment by sdogruyol** [comment]
@denysvitali can you add specs for this?

**Comment by denysvitali** [comment]
@sdogruyol Yes, I'll do my best as soon as I can

**Comment by denysvitali** [comment]
it "should work" now 👍

File: src/kemal/static_file_handler.cr [hunk 1]
```diff
@@ -76,6 +79,65 @@ module Kemal
...
+              skipped += 1024
+            end
+            if skipped - startb > 0
+              file.skip(skipped - startb)
+            end
+          else
+            file.skip(startb)
+          end
+        rescue ex
+          puts "#{ex.message} #{ex.backtrace}"
```
Comment (Sija) [inline src/kemal/static_file_handler.cr]: debug leftovers?

File: src/kemal/static_file_handler.cr [hunk 2]
```diff
@@ -76,6 +79,65 @@ module Kemal
...
+        end
+
+        begin
+          realendbytes = endb
+          IO.copy(file, env.response, realendbytes - startb)
+        rescue ex
+          puts "Handled ex: #{ex.message}"
```
Comment (Sija) [inline src/kemal/static_file_handler.cr]: ditto

File: .gitignore [hunk 3]
```diff
@@ -6,3 +6,5 @@
 # Libraries don't need dependency lock
 # Dependencies will be locked in application that uses them
 /shard.lock
+samples/multipart_file.cr
+/static/
```
Comment (Sija) [inline .gitignore]: Modifying `.gitignore` file seems out of scope of this PR.

File: spec/static_file_handler.cr [hunk 4]
```diff
@@ -99,4 +101,33 @@ describe Kemal::StaticFileHandler do
...
+    %w(GET).each do |method|
+      headers = HTTP::Headers{"Range" => "0-100"}
+      response = handle HTTP::Request.new(method, "/dir/test.txt", headers)
+      response.status_code.should eq(206 || 200)
+      if response.status_code == 206
+        response.headers.has_key?("Content-Range").should eq true
+        match = response.headers["Content-Range"].match(/bytes (\d+)-(\d+)\/(\d+)/)
+        match.should_not be nil
+        if match
+          startRange = match[1].to_i {0}
```
Comment (Sija) [inline spec/static_file_handler.cr]: Why use `CamelCase` here? Lets stick to `snake_case` convention for naming variables.

**Comment by denysvitali** [comment]
Thanks for the review, I agree with all of your comments. I'll fix my code ASAP and commit the changes

File: src/kemal/static_file_handler.cr (hunk 1 above)
Comment (Sija) [inline src/kemal/static_file_handler.cr]: this should raise and let it through to the exception handler i think :)

File: src/kemal/static_file_handler.cr [hunk 5]
```diff
@@ -76,6 +79,65 @@ module Kemal
...
+      end
+
+      if endb == 0
+        endb = fileb
+      end
+
+      if startb < endb && endb <= fileb
+        env.response.status_code = 206
+        env.response.content_length = endb - startb
+        env.response.headers.add("Content-Range", "bytes #{startb}-#{endb - 1}/#{fileb}") # MUST
```
Comment (Sija) [inline src/kemal/static_file_handler.cr]: is there any particular reason you're calling `#add` here, instead of `#[]=` (like for `X-Streaming` below)? i'm just curious :)

File: src/kemal/static_file_handler.cr (hunk 1 above)
Comment (denysvitali) [inline src/kemal/static_file_handler.cr]: Agree, will fix 👍

File: src/kemal/static_file_handler.cr [hunk 6]
```diff
@@ -76,6 +79,65 @@ module Kemal
...
+        env.response.headers["X-Streaming"] = "true" #Not in RFC, but useful for debug
```
Comment (Sija) [inline src/kemal/static_file_handler.cr]: Are there any usecases for adding non-standard header like this?

Comment (denysvitali) [inline src/kemal/static_file_handler.cr]: None, I'll remove it, since Content-Range is added only when a range is sent.

File: src/kemal/static_file_handler.cr (hunk 5 above)
Comment (denysvitali) [inline src/kemal/static_file_handler.cr]: Nope, I'll fix this inconsistency

File: spec/static_file_handler.cr (hunk 4 above)
Comment (denysvitali) [inline spec/static_file_handler.cr]: Agree

**Comment by denysvitali** [comment]
Code fixed as per code review, enjoy!

**Comment by sdogruyol** [comment]
Thank you @denysvitali 👍

**Comment by denysvitali** [comment]
We may have to fix it soon or later, apparently it doesn't fully work properly (I may have missed something I think).  
I will test as soon as I can, file a bug report and propose a fix

**Comment by sdogruyol** [comment]
@denysvitali what's not working properly?

**Comment by denysvitali** [comment]
I have a bug that may be related to #300 , and I'm trying to investigate the reason behind the Broken pipe (it may be my Firefox Developer installation)

**Comment by sdogruyol** [comment]
@denysvitali i'll also try to reproduce

**Comment by sdogruyol** [comment]
I can reproduce this with stress loading serving big static files like videos.

Using `wrk` 

```
wrk -c 10 -d 20 http://localhost:3000/video.mp4
```

Kemal process crashes when the test finishes (not in between)
```
Unhandled exception in spawn:
Error writing file: Broken pipe (Errno)
0x106fff355: *CallStack::new:CallStack at ??
0x107001c4e: *raise<Errno>:NoReturn at ??
0x10706bbb9: *TCPSocket+@IO::Buffered#flush:TCPSocket+ at ??
0x10707345c: *TCPSocket+@IO::Buffered#close:Nil at ??
0x10706fe54: *HTTP::Server::RequestProcessor#process<(OpenSSL::SSL::Socket::Server | TCPSocket+), (OpenSSL::SSL::Socket::Server | TCPSocket+), IO::FileDescriptor>:Nil at ??
0x10700367c: *Fiber#run:(IO::FileDescriptor | Nil) at ??

Invalid memory access (signal 10) at address 0x109a47520
[4412989111] *CallStack::print_backtrace:Int32 +39
[4412495875] __crystal_sigfault_handler +35
[140735702649786] _sigtramp +26
```

//cc @asterite

**Comment by denysvitali** [comment]
Exactly, that's the issue. For an application of mine I needed to serve many static files in a single page, and I was always having a broken pipe problem. I then swtiched my static files to nginx, and hotfixed the problem.
What can it be? Is it kemal related or Crystal related?

At first I thought it was a Firefox problem, because when a pipe is broken is mostly caused by the browser requesting to close the connection, but apparently with `wrk` this isn't the case

**Comment by sdogruyol** [comment]
Let's discuss on #300
